crates/sui-framework/packages/bridge/sources/limiter.move

```diff
@@ -0,0 +1,192 @@
+
+module bridge::limiter {
+
+    use std::type_name;
+    use std::type_name::TypeName;
+    use std::vector;
+    use sui::clock;
+    use sui::clock::Clock;
+    use sui::math::pow;
+    use sui::vec_map;
+    use sui::vec_map::VecMap;
+    use bridge::treasury;
+    use bridge::chain_ids::BridgeRoute;
+    #[test_only]
+    use sui::test_scenario;
+    #[test_only]
+    use sui::test_utils::destroy;
+    #[test_only]
+    use bridge::chain_ids;
+    #[test_only]
+    use bridge::eth::ETH;
+
+    const ETransferAmountExceedLimit: u64 = 0;
+
+    // TODO: Make this configurable?
+    const DEFAULT_TRANSFER_LIMIT: u64 = 0;
+
+    struct TransferLimiter has store {
+        transfer_limits: VecMap<BridgeRoute, u64>,
+        notional_values: VecMap<TypeName, u64>,
+        // Per hour transfer amount for each bridge route
+        transfer_records: VecMap<BridgeRoute, TransferRecord>,
+    }
+
+    struct TransferRecord has store {
+        last_recorded_hour: u64,
+        per_hour_amounts: vector<u64>,
+        total_amount: u64
+    }
+
+    public fun new(transfer_limits: VecMap<BridgeRoute, u64>): TransferLimiter {
+        TransferLimiter {
+            transfer_limits,
+            notional_values: vec_map::empty(),
+            transfer_records: vec_map::empty()
+        }
+    }
+
+    // Current hour since unix epoch
+    fun current_hour_since_epoch(clock: &Clock): u64 {
+        clock::timestamp_ms(clock) / 3600000
+    }
+
+    public fun check_and_record_transfer<T>(
+        clock: & Clock,
+        self: &mut TransferLimiter,
+        route: BridgeRoute,
+        amount: u64
+    ) {
+        // Create record for route if not exists
+        if (!vec_map::contains(&self.transfer_records, &route)) {
+            vec_map::insert(&mut self.transfer_records, route, TransferRecord {
+                last_recorded_hour: 0,
+                per_hour_amounts: vector[],
+                total_amount: 0
+            })
+        };
+        let record = vec_map::get_mut(&mut self.transfer_records, &route);
+        let current_hour_since_epoch = current_hour_since_epoch(clock);
+
+        // First clean up old transfer histories
+        let i = vector::length(&record.per_hour_amounts);
+        while (i > 0 && record.last_recorded_hour < (current_hour_since_epoch - 24 + i)) {
+            record.total_amount = record.total_amount - vector::remove(&mut record.per_hour_amounts, 0);
+            i = i - 1;
+        };
+        // Backfill missing hours
+        while (record.last_recorded_hour < current_hour_since_epoch) {
+            // only backfill up to 24 hours
+            if (record.last_recorded_hour < current_hour_since_epoch - 23) {
+                record.last_recorded_hour = current_hour_since_epoch - 23;
+            }else {
+                record.last_recorded_hour = record.last_recorded_hour + 1;
+            };
+            vector::push_back(&mut record.per_hour_amounts, 0);
+        };
+
```

Comment:
This logic is quite complicated to my little brain to reason about. How do we feel about using maps (dynamic fields) to store the values? I hope storage fee is negligible in this case for better code readability

Reply:
or alternative can i suggest the following changes:
1. add `hour_tail` and `hour_head` to represent the beginning and end hour currently in record. 
1. add `fun remove_stale_hour_maybe(record)` which looks like
```
fun remove_stale_hour_maybe(record) {
   // remove tails until it's within 24 hours range
   while record.hour_tail + 23 < current_hour_since_epoch && record.hour_tail < record.hour_head {
      record.total_amount = record.total_amount - vector::remove(&mut record.per_hour_amounts, 0);
      record.hour_tail = record.hour_tail + 1;
   }
}
```

2. add `fun append_new_empty_hours(record, current_hour_since_epoch)`
```
    fun append_new_empty_hours(record, current_hour_since_epoch) {
        if record.hour_head = current_hour_since_epoch {
            return; // nothing to backfill
        }

        // If tail is even older than 24 hours ago, advance it to that.
        let target_tail = current_hour_since_epoch - 23;
        if record.hour_tail < target_tail {
             record.hour_tail = target_tail;
        }
       
        // If old head is even older than target tail, advance it to that.
        if record.hour_head < target_tail {
             record.hour_head = target_tail;
        }

        // Backfill from head to current hour
        while record.hour_head < current_hour_since_epoch {
           record.hour_head = record.hour_head + 1;
           vector::push_back(&mut record.per_hour_amounts, 0);
        } 
    }
```